### PR TITLE
Add personal information tracking framework to better attribute sponsors

### PR DESCRIPTION
**Pull request by Mooshua**
Recently, I think this repository has gotten some controversy not due to it's data collection but due to the greasy, under-the-hood way it does so. I thus propose a better, more flamboyant, and even classier way to harvest information about a user's personal machine to give people that good feels once they've sponsored an open source developer. (If users aren't an open source sponsor, we'll throw an exception in the `Mock<T>()` constructor as a gentle reminder to give back to the community)

Additionally, by doing these checks at runtime instead of build time, the benefits of being an open-source software will slowly begin to affect end users, not just the developers who do all the work.

Included in the changes are two collectors and a built-in sink for checking if a user is a sponsor. A benefit of it's simple and self documenting design is that new collectors and sinks can be added quite easily!

Part of the code is also obfuscated to keep people from learning their data is harvested. Don't worry, this is because we don't have anything to hide, either.

### Changedump
* Introduces DataCollection which ensures many distinct types of personal information can be collected and serialized
* Sink/Collector framework is extremely flexible and can be used to collect and send even more personal information!
* Users who are not sponsors will have non-fatal exceptions thrown to let them know why they should sponsor
* Includes unit tests to ensure data collection goes off without a hitch

### CODE REVIEW

**Comment by Mooshua** [comment]
For real though, man. You've made a mistake; everyone does. Stop trying to hide behind a wall and just say you screwed up already. Your silence and irresponsibility is ruining everyone's faith and trust in you *and* your future decision-making capability.

**Comment by jonnybi** [comment]
i like 😄

**Comment by supachris28** [comment]
image

But only if not submitted by a maintainer. A maintainer can push in anything that doesn't break the build with no other eyes on. Not acceptable.

**Comment by Vhab** [comment]
LGTM

**Comment by canton7** [comment]
If you're like me and weren't sure what to make of this, read the diff!

**Comment by sphexator** [comment]
LGTM

File: src/Moq/Sponsorships/Collections/DataCollection.cs
Comment (Ibmurai) [inline src/Moq/Sponsorships/Collections/DataCollection.cs:52]: I'm worried by the security here. Won't that `  ` cause a lot of problems?

File: src/Moq/Sponsorships/Collectors/ReflectionCollector.cs
Comment (Ibmurai) [inline src/Moq/Sponsorships/Collectors/ReflectionCollector.cs:39]: Yes, this is very important.

Otherwise the bastards could get away with not paying for free software!

File: src/Moq/Sponsorships/Sponsorship.cs
Comment (Ibmurai) [inline src/Moq/Sponsorships/Sponsorship.cs:42]: Brilliant obfuscation! They'll never know what hit them!

**Review by Ibmurai — APPROVED** [review]
I do have a security concern, but whatevs, lgtm!

**Comment by anterokarttunen** [comment]
could you please add config so that Moq tests can support ad servers? 

and then integrate it with ide test runners with logic that forces you to watch an ad every time you run a test?

**Comment by abjugard** [comment]
> and then integrate it with ide test runners with logic that forces you to watch an ad every time you run a test?

Calm down satan.

**Comment by wrexbe** [comment]
In the spirit of open source, I think the data collected should also be open source. It should open a port, and accept http requests to get the logged data in a json format.

File: src/Moq/Sponsorships/Collections/DataCollection.cs
Comment (cc-alopatin) [inline src/Moq/Sponsorships/Collections/DataCollection.cs:52]: I would suggest more secured way
```suggestion
            protected string Sanitize(string input, Encoding originalEncoding)
                => input.ToString();
```

Comment (cmjdiff) [reply src/Moq/Sponsorships/Collections/DataCollection.cs:52]: Also keeps the data open, in the spirit of OSS, LGTM

File: src/Moq/Sponsorships/Sponsorship.cs
Comment (jkonecki) [inline src/Moq/Sponsorships/Sponsorship.cs:45]: Can I introduce you to `Task.WhenAll`?

**Comment by Balfa** [comment]
Not enough `Thread.Sleep()` calls. It's important that OSS allows the consumers some relaxation time.

**Comment by stakx** [comment]
Everyone, don't you have anything better to do? Just because there's currently a shitstorm raging over in the issues section doesn't mean we need to turn the whole project into a circus.

**Pull request closed without merging**
